knodle/transformation/majority.py

```diff
@@ -100,115 +129,273 @@ def z_t_matrices_to_majority_vote_probs(
-        # convert labels represented as a prob distribution to a single label using majority voting
-        if multi_label:
-            kwargs = {"choose_random_label": True, "other_class_id": other_class_id, "threshold": multi_label_threshold}
-            noisy_y_train = np.apply_along_axis(probabilities_to_binary_multi_labels, axis=1, arr=noisy_y_train, **kwargs)
-        else:
-            kwargs = {"choose_random_label": True, "other_class_id": other_class_id}
-            noisy_y_train = np.apply_along_axis(probabilities_to_majority_vote, axis=1, arr=noisy_y_train, **kwargs)
 
-    return model_input_x, noisy_y_train, rule_matches_z
+def probabilities_to_majority_vote(probs: np.ndarray, ties_strategy: str = "random", other_class_id: int = None) -> int:
+    """Transforms a vector of probabilities to its majority vote. If there is one class with clear majority, return it.
+    If there are more than one class with equal probabilities: either select one of the classes randomly or assign to
+    the sample the other class id.
+    Args:
+        probs: Vector of probabilities for 1 sample. Shape: classes x 1
+        ties_strategy: what to do with ties (choose a random label or an other class)
+        other_class_id: Class ID being used, if there's no clear majority.
+    Returns: An array of classes.
+    """
+
+    row_max = np.max(probs)
+    num_occurrences = (row_max == probs).sum()
+    if num_occurrences == 1:
+        return int(np.argmax(probs))
+    elif ties_strategy == "other":
+        return other_class_id
+    elif ties_strategy == "random":
+        max_ids = np.where(probs == row_max)[0]
+        return int(np.random.choice(max_ids))
+    else:
+        raise ValueError("Specify how to resolve unclear majority votes.")
 
 
 def probabilities_to_binary_multi_labels(
-        probs: np.ndarray, choose_random_label: bool = True, other_class_id: int = None, threshold: float = 0.5
+        probs: np.ndarray, ties_strategy: str = "random", other_class_id: int = None, threshold: float = 0.5
 ) -> np.ndarray:
     """
     probs: Vector of probabilities for 1 sample. Shape: classes x 1
     choose_random_label: Choose a random label, if there's no clear majority.
     other_class_id: Class ID being used, if there's no clear majority
-    multi_label: boolean value, whether the classification is multi-label
     threshold: a value for calculation the classes in case of multi-label classification: if a class has
         a probability greater than the threshold, this class will be selected as a true one
     """
     probs[probs >= threshold] = 1
     probs[probs < threshold] = 0
 
     if np.all((probs == 0)):
-        if choose_random_label:
+        if ties_strategy == "random":
             probs[:, random.randrange(probs.shape[1])] = 1
-        elif other_class_id is not None:
+        elif ties_strategy == "other":
             probs[:, other_class_id] = 1
-
     return probs
 
+
+def check_input_validity(*inputs) -> None:
```

Comment:
I would do these checks where they are needed. Like unmatched strategy at handle unlabeled. I would also call both unlabeled or both umatched.

Reply:
Then it'd be also easier to write Unit tests.